CRDT PN counter

mmedenjak opened the pull request:
Adds the PN-Counter (Positive-Negative Counter) implementation together with the infrastructure common for all CRDTs. 

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/1912

Depends on these PRs being merged first: 
https://github.com/hazelcast/hazelcast-client-protocol/pull/113
https://github.com/hazelcast/hazelcast/pull/12267

Danny-Hazelcast commented:
given a rebase and if all comments have been addressed.
It would be very help full to have this RP merged, 
this would allow the QA team to start testing, earlier the better from QA point of view

metanet commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,399 @@
+        try {
+            return invokeOnAddress(request, target);
+        } catch (ConsistencyLostException e) {
+            if (excludedAddresses == null) {
```
should we print a log here?

mmedenjak replied:
We can. It's a proxy so there may be lots of log messages (from different proxies, from concurrent invocations) but OTOH it should happen rarely and it may make sense. WDYT?

metanet replied:
I see. I think it would be still good to print a log.

mmedenjak replied:
Agree, will do.

mmedenjak commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,399 @@
+     * @param excludedAddresses the addresses to exclude
+     * @return list of possible CRDT replica addresses
+     */
+    private List<Address> getReplicaAddresses(ArrayList<Address> excludedAddresses) {
```
I wanted to avoid creating a new list for every invocation of every get or update method, even in the case where we always go to the same data member. WDYT?

metanet replied:
maybe we can create a static empty list and pass it initially instead of a null value.

mmedenjak replied:
Thanks, will do.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/AbstractCRDTReplicationOperation.java:
```diff
@@ -0,0 +1,98 @@
+import static com.hazelcast.util.MapUtil.createHashMap;
+
+/**
+ * Base class for CRDT replication operations. It support replicating a
```
minor typo: `it supports replicating...`

mmedenjak replied:
👍

metanet commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,399 @@
+            AtomicReferenceFieldUpdater.newUpdater(ClientPNCounterProxy.class, VectorClock.class, "observedClock");
+    private volatile Address currentTargetReplicaAddress;
+    private final Object targetSelectionMutex = new Object();
+    private final Object maxConfiguredReplicaCountMutex = new Object();
```
I think we can survive without this mutex. even if max configured replica count is fetched from hz servers multiple times, they will all return the same value.

mmedenjak replied:
True. I wanted to avoid different client threads retrieving the same config but perhaps it's not such an issue.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/CRDT.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @param <T>
+ */
+public interface CRDT<T extends CRDT<T>> {
```
sexy 💯

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/CRDTMigrationTask.java:
```diff
@@ -0,0 +1,132 @@
+                return;
+            }
+            final Collection<Member> members = nodeEngine.getClusterService().getMembers(DATA_MEMBER_SELECTOR);
+            final Member firstDataMember = members.iterator().next();
```
can we add a comment here to describe the reasoning? If I am the first data member, there is no other member to migrate my data..

mmedenjak replied:
Will do. IMO this is the biggest downside of the member list approach - the first member owns everything. The downside of the partition table based approach being that there is max 7 replicas. Maybe we can find an even better approach in the future.

mmedenjak commented:
@metanet addressed all review comments
@sancar changed the client proxy to assume a non-null initial observed vector clock

metanet commented:
Thanks. I am still reviewing the changes on my machine.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounter.java:
```diff
@@ -0,0 +1,216 @@
+ * If no new updates are made to the shared state, all nodes that can
+ * communicate will eventually have the same data.
+ * <p>
+ * Each replica is identified by a UUID. This identifier should be
```
do we need to describe the uuid logic here? We don't expose it in the interface at all and I think it is an internal concern.

mmedenjak replied:
👍

mmedenjak replied:
Removed

taburet commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,451 @@
+     *                          counter value before the addition, {@code false}
+     *                          if it should return the value after the addition
+     * @param excludedAddresses the addresses to exclude when choosing a replica
+     *                          address, may be {@code null}
```
Seems like it can't be `null` actually, this will produce NPE.

mmedenjak replied:
Leftover, thanks

mmedenjak replied:
Fixed.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounter.java:
```diff
@@ -0,0 +1,216 @@
+ * When invoking updates from non-replica instance, the invocation is remote.
+ * This may lead to indeterminate state - the update may be applied but the
+ * response has not been received. In this case, the caller will be notified
+ * with a {@link com.hazelcast.spi.exception.TargetDisconnectedException}.
```
I think the invocation system can also throw `MemberLeftException` if the invocation is between hz instances.

mmedenjak replied:
True, will update.

mmedenjak replied:
Added comment

taburet commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,451 @@
+        } catch (ConsistencyLostException e) {
+            logger.fine("Unable to provide session guarantees when sending operations to " + target
+                    + ", choosing different target");
+            if (excludedAddresses == EMPTY_ADDRESS_LIST) {
```
It may be a good idea to shrink these 4 repetitive lines below into a single utility method, this will raise readability, but this's just a matter of taste.

mmedenjak replied:
I was thinking about that, seemed like a small gain for the extra method but I'll try again

mmedenjak replied:
Extracted the method, as well as for the `ClientPNCounterProxy` and the `PNCounterProxy`

taburet commented on hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java:
```diff
@@ -0,0 +1,451 @@
+     * if the {@code lastException} is {@code null}.
+     *
+     * @param excludedAddresses the addresses to exclude when choosing a replica
+     *                          address, may be {@code null}
```
Can't be `null`.

mmedenjak replied:
👍

mmedenjak replied:
Fixed.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/MutationDisallowedException.java:
```diff
@@ -0,0 +1,38 @@
+ * process of cleaning up</li>
+ * <li>a target replica is shutting down and in the process of replicating
+ * unreplicated CRDT state</li>
+ * <li>there are no data members to store the state</li>
```
I think this exception is a little bit confusing. For instance, in the first and second cases, where a node does not perform an operation because it is no longer a crdt replica, we retry the operation in other nodes. In these two cases, this exception is handled internally. In the last case, there is no data node in the cluster so we propagate the exception to the caller. We already have an exception for this case: `NoDataMemberInClusterException`. Can we throw `NoDataMemberInClusterException` to the caller and keep this exception internal? WDYT?

mmedenjak replied:
Didn't know about that exception. I will update it. OTOH the `MutationDisallowedException` may still be thrown if it is the last exception and the client has no other replicas it can try. Not sure how to avoid this one.

metanet replied:
Do we have a case where there is at least one data member but still cannot find a crdt replica?

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounter.java:
```diff
@@ -0,0 +1,216 @@
+     *                                                     members
+     * @throws UnsupportedOperationException               if the cluster version
+     *                                                     is less than 3.10
+     * @throws com.hazelcast.crdt.ConsistencyLostException if the session guarantees
```
minor comment: we can import `ConsistencyLostException`

mmedenjak replied:
Fixed.

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounterProxy.java:
```diff
@@ -0,0 +1,319 @@
+            throw new UnsupportedOperationException(
+                    "PNCounter operations are not supported when cluster version is less than 3.10");
+        }
+        operation.setValidateTarget(false);
```
afaik, target validation is only done for partition operations. We don't need to set this flag since we send operations to specific members here.

mmedenjak replied:
Fixed.

mmedenjak commented:
@metanet @taburet addressed all review comments except for the one for having a different exception. I'll do that now. I also fixed the state clear in the `PNCounterService#reset` method (previously the state was cleared during split brain merging)

metanet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounter.java:
```diff
@@ -0,0 +1,197 @@
+ * guarantees. These guarantees are session guarantees which means that if
+ * no replica with the previously observed state is reachable, the session
+ * guarantees are lost and the method invocation will throw a
+ * {@link ConsistencyLostException}. This does not mean
```
I really like this `ConsistencyLostException` part. good work on this.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/crdt/VectorClock.java:
```diff
@@ -0,0 +1,129 @@
+ * <p>
+ * See https://en.wikipedia.org/wiki/Vector_clock
+ */
+public class VectorClock implements DataSerializable {
```
Changed.

taburet commented on hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounterService.java:
```diff
@@ -0,0 +1,256 @@
+            final String counterName = vectorClockEntry.getKey();
+            final VectorClock vectorClock = vectorClockEntry.getValue();
+            final PNCounterImpl counter = counters.get(counterName);
+            if (counter == null) {
```
Seems like this is possible only when someone destroyed the counter while we are trying to migrate it. In this case it may survive on another member because it may be migrated after the destroy operation is already performed on that member. If I understand things correctly, this may lead to a formation of flaky counters which are present only on some nodes.

mmedenjak replied:
That's one way. One other way would be that there are two concurrent migration operations - e.g. there was a merge followed by an membership event on the same replica. These will start two migration tasks but only one can remove a counter from the map.

As for the concurrent destroy, true, the counter may survive on one node. In that case, it will be again replicated to other members, since from the replication service we on the next replication round see that other members haven't observed the counter. So in a sense, the counter survived the destroy.

taburet replied:
Thank you for the clarification. This destroy issue doesn't look that critical unless one relies heavily on the existence of counters or "remaps" their names so they change the semantics of counters. Seems like those are not frequent use cases.

mmedenjak commented:
@metanet @taburet @sancar 
Addressed all comments and:
- added split brain tests with different replica counts
- added consistency lost exception tests for client and lite member
- added test for no data member in cluster
- the proxies now throw `NoDataMemberException` when there are no data members
- moved the `VectorClock` to `com.hazelcast.cluster.impl` and `ConsistencyLostException` to `com.hazelcast.core` so they can easily be reused with different services
- the proxies now catch `HazelcastException` on the invocations instead of separate `ConsistencyLostException`, `TargetNotReplicaException`, ... The reason for this was that the client and member were throwing various exceptions when the replica disappears but from the point of the proxy the exact exception is irrelevant - we should just log the exception and try with a different replica.

mmedenjak commented:
run-lab-run

mmedenjak commented:
run-lab-run
